File: src/ta_foundation/strategies/shared/PantheonStopEngine.cs
/*
 * ═══════════════════════════════════════════════════════════════════════════════
 *  PantheonStopEngine — pure stop/trail math shared by Strategy + AddOn + tests
 * ═══════════════════════════════════════════════════════════════════════════════
 *
 *  WHY THIS FILE EXISTS
 *  ─────────────────────────────────────────────────────────────────────────────
 *  The stop-in-profit / trail logic used to live inside PantheonMaster (a Strategy),
 *  so it could only be exercised through a live or Playback Strategy — one position
 *  at a time. This class lifts that math out into plain C# with ZERO NinjaTrader
 *  dependencies, so the identical logic can be driven by:
 *    1. PantheonMaster (live path delegates to it),
 *    2. the TaFoundation AddOn live-test harness (Account-API order driver),
 *    3. a C# unit test / battery harness (no NT at all),
 *    4. and mirrored 1:1 by the Python replica
 *       (src/ta_foundation/analysis/exits/nt_atr_trail_parity.py).
 *
 *  BUILD / LINKING
 *  ─────────────────────────────────────────────────────────────────────────────
 *  Because it has no NT references, this single source file is LINKED into both
 *  the external Visual Studio AddOn project (<Compile Include="..\..\shared\
 *  PantheonStopEngine.cs" />) and dropped into NinjaScript custom for the Strategy.
 *  One source of truth — do not fork it.
 *
 *  FAITHFULNESS
 *  ─────────────────────────────────────────────────────────────────────────────
 *  Every branch mirrors PantheonMaster.cs exactly (line refs noted per method as of
 *  the extraction). The engine is PURE: callers own the running state (favorable
 *  extreme, peak open profit, breakEvenActive, lastSubmittedStop) and the live
 *  side-of-market guard + actual order submission. The engine only answers
 *  "given this state, what stop price does the policy want, and does it improve?"
 * ═══════════════════════════════════════════════════════════════════════════════
 */

using System;

namespace TaFoundation.Pantheon
{
	/// <summary>
	/// Exit policy selector. Distinct from the Strategy's NinjaScript
	/// <c>PantheonExitPolicy</c> enum (which is serialized into workspaces) so this
	/// shared class carries no NinjaScript coupling. Callers map their enum to this.
	/// </summary>
	public enum PantheonStopPolicy
	{
		FixedRR       = 0,
		AtrTrail      = 1,
		Chandelier    = 2,
		Giveback      = 3,
		FixedStop     = 4,
		BreakEvenOnly = 5
	}

	/// <summary>Immutable per-trade configuration (mirror of the strategy's Discovery Exit params).</summary>
	public struct PantheonStopConfig
	{
		public PantheonStopPolicy Policy;
		public double TickSize;
		public double PointValue;

		public int    StopTicks;             // initial protective stop distance (ticks)
		public int    TargetTicks;           // FixedRR target distance (ticks)
		public double AtrTrailMultiple;      // AtrTrail / Chandelier ATR multiple
		public int    ChandelierLookback;    // bars (caller supplies the ref price)
		public double GivebackPct;           // fraction of peak open profit to give back
		public int    BreakEvenTriggerTicks; // profit (ticks) before BE arms
		public int    BreakEvenPlusTicks;    // BE stop offset above/below entry (ticks)
	}

	/// <summary>Result of a trail computation. <see cref="HasProposal"/> false means leave the stop where it is.</summary>
	public struct PantheonTrailResult
	{
		public bool   HasProposal;        // a (rounded) proposed stop price is present
		public double ProposedStop;       // rounded to tick; only valid when HasProposal
		public bool   BreakEvenActivated; // BreakEvenOnly: true on the tick BE first arms
	}

	/// <summary>
	/// Pure stop/trail math. All methods are static and side-effect free; the caller
	/// owns and threads the running state between calls.
	/// </summary>
	public static class PantheonStopEngine
	{
		public const int Long  =  1;
		public const int Short = -1;

		/// <summary>Round a price to the instrument tick grid (mirror of RoundToTickSize).</summary>
		public static double RoundToTick(double price, double tickSize)
		{
			if (tickSize <= 0.0) return price;
			return Math.Round(price / tickSize) * tickSize;
		}

		/// <summary>
		/// Initial protective stop price. Mirror of PantheonMaster.GetInitialStopPrice
		/// (discovery path): StopTicks below entry for long / above for short.
		/// </summary>
		public static double InitialStop(int dir, double entryPrice, PantheonStopConfig cfg)
		{
			return RoundToTick(entryPrice - dir * cfg.StopTicks * cfg.TickSize, cfg.TickSize);
		}

		/// <summary>True only for FixedRR — the one discovery policy that carries a target (ShouldUseTarget).</summary>
		public static bool UsesTarget(PantheonStopConfig cfg)
		{
			return cfg.Policy == PantheonStopPolicy.FixedRR;
		}

		/// <summary>Initial target price (FixedRR). Mirror of GetInitialTargetPrice (discovery path).</summary>
		public static double InitialTarget(int dir, double entryPrice, PantheonStopConfig cfg)
		{
			return RoundToTick(entryPrice + dir * cfg.TargetTicks * cfg.TickSize, cfg.TickSize);
		}

		/// <summary>
		/// Ratchet test. Mirror of MoveStopIfImproved / MoveHistoricalStopIfImproved:
		/// a long stop must rise &gt; 0.5 tick above the last stop; a short must fall the same.
		/// </summary>
		public static bool Improves(int dir, double proposedStop, double lastStop, double tickSize)
		{
			if (dir == Long)  return proposedStop > lastStop + tickSize * 0.5;
			return proposedStop < lastStop - tickSize * 0.5;
		}

		/// <summary>Open profit in account currency at <paramref name="price"/>. Mirror of GetOpenProfitCurrencyFromPrice.</summary>
		public static double OpenProfitCurrency(int dir, double price, double entryPrice, double pointValue, int quantity)
		{
			if (quantity <= 0) return 0.0;
			double points = dir == Long ? price - entryPrice : entryPrice - price;
			return points * pointValue * quantity;
		}

		/// <summary>
		/// Core trail computation — the exact per-policy switch from
		/// PantheonMaster.ManageLiveDynamicStop, with NO NinjaScript dependency.
		///
		/// The caller maintains and passes the running state:
		///   favorableExtreme       — running max(price) for a long / running min for a short,
		///                            updated tick-by-tick (live) or bar-by-bar (backtest).
		///   peakOpenProfitCurrency — running max open profit ($), for Giveback.
		///   chandelierRefPrice     — long: highest High over ChandelierLookback bars;
		///                            short: lowest Low. Pass 0 if not using Chandelier.
		///   breakEvenActive        — whether BreakEvenOnly has already armed this trade.
		///
		/// Returns a (rounded) proposed stop when the policy wants one; the caller still
		/// applies <see cref="Improves"/> (ratchet) and its own side-of-market guard before
		/// submitting. For BreakEvenOnly, <see cref="PantheonTrailResult.BreakEvenActivated"/>
		/// tells the caller to flip its breakEvenActive flag.
		/// </summary>
		public static PantheonTrailResult ComputeTrailStop(
			int dir,
			double entryPrice,
			double triggerPrice,
			double favorableExtreme,
			double currentAtr,
			double peakOpenProfitCurrency,
			int quantity,
			double chandelierRefPrice,
			bool breakEvenActive,
			PantheonStopConfig cfg)
		{
			var result = new PantheonTrailResult { HasProposal = false, ProposedStop = 0.0, BreakEvenActivated = false };
			double tick = cfg.TickSize;

			switch (cfg.Policy)
			{
				case PantheonStopPolicy.BreakEvenOnly:
				{
					if (breakEvenActive) break;
					double triggerLevel = entryPrice + dir * cfg.BreakEvenTriggerTicks * tick;
					bool armed = dir == Long ? triggerPrice >= triggerLevel : triggerPrice <= triggerLevel;
					if (armed)
					{
						result.BreakEvenActivated = true;
						result.HasProposal  = true;
						result.ProposedStop = RoundToTick(entryPrice + dir * cfg.BreakEvenPlusTicks * tick, tick);
					}
					break;
				}

				case PantheonStopPolicy.AtrTrail:
				{
					double proposed = favorableExtreme - dir * cfg.AtrTrailMultiple * currentAtr;
					result.HasProposal  = true;
					result.ProposedStop = RoundToTick(proposed, tick);
					break;
				}

				case PantheonStopPolicy.Chandelier:
				{
					// long: pull from the higher of (lookback high, favorable extreme); short: the lower.
					double refPrice = dir == Long
						? Math.Max(chandelierRefPrice, favorableExtreme)
						: Math.Min(chandelierRefPrice, favorableExtreme);
					double proposed = refPrice - dir * cfg.AtrTrailMultiple * currentAtr;
					result.HasProposal  = true;
					result.ProposedStop = RoundToTick(proposed, tick);
					break;
				}

				case PantheonStopPolicy.Giveback:
				{
					if (peakOpenProfitCurrency > 0.0)
					{
						double protectedProfit = peakOpenProfitCurrency * (1.0 - cfg.GivebackPct);
						double proposed = StopPriceFromProtectedProfit(dir, entryPrice, protectedProfit, cfg.PointValue, quantity);
						result.HasProposal  = true;
						result.ProposedStop = RoundToTick(proposed, tick);
					}
					break;
				}

				// FixedRR / FixedStop: the initial stop never trails.
				case PantheonStopPolicy.FixedRR:
				case PantheonStopPolicy.FixedStop:
				default:
					break;
			}

			return result;
		}

		/// <summary>Mirror of GetStopPriceFromProtectedOpenProfit.</summary>
		public static double StopPriceFromProtectedProfit(int dir, double entryPrice, double protectedProfitCurrency, double pointValue, int quantity)
		{
			double denom = pointValue * Math.Max(1, quantity);
			double protectedPoints = denom == 0.0 ? 0.0 : protectedProfitCurrency / denom;
			return dir == Long ? entryPrice + protectedPoints : entryPrice - protectedPoints;
		}
	}
}

File: src/ta_foundation/analysis/exits/pantheon_stop_engine.py
"""Python mirror of ``strategies/shared/PantheonStopEngine.cs``.

This is a 1:1 port of the pure C# stop/trail math so the live NinjaTrader path
(PantheonMaster + the AddOn harness) and the fast Python validation battery
exercise the *same* logic. Every function maps to a static method in
``PantheonStopEngine``; keep the two in lockstep — if you change a branch here,
change it there (and vice versa).

The companion driver/battery lives in ``pantheon_trail_battery.py``; the original
AtrTrail-only vectorized replica in ``nt_atr_trail_parity.py`` remains the
backtest-parity reference and is cross-checked against this engine in tests.
"""
from __future__ import annotations

from dataclasses import dataclass
from enum import Enum

LONG = 1
SHORT = -1


class PantheonStopPolicy(str, Enum):
    FIXED_RR = "FixedRR"
    ATR_TRAIL = "AtrTrail"
    CHANDELIER = "Chandelier"
    GIVEBACK = "Giveback"
    FIXED_STOP = "FixedStop"
    BREAK_EVEN_ONLY = "BreakEvenOnly"


@dataclass(frozen=True)
class PantheonStopConfig:
    """Mirror of the C# ``PantheonStopConfig`` (= the strategy's Discovery Exit params)."""
    policy: PantheonStopPolicy
    tick_size: float = 0.25
    point_value: float = 20.0          # NQ
    stop_ticks: int = 60
    target_ticks: int = 90
    atr_trail_multiple: float = 2.0
    chandelier_lookback: int = 20
    giveback_pct: float = 0.40
    breakeven_trigger_ticks: int = 30
    breakeven_plus_ticks: int = 4


@dataclass(frozen=True)
class TrailResult:
    has_proposal: bool = False
    proposed_stop: float = 0.0
    breakeven_activated: bool = False


def round_to_tick(price: float, tick_size: float) -> float:
    if tick_size <= 0.0:
        return price
    return round(price / tick_size) * tick_size


def initial_stop(direction: int, entry_price: float, cfg: PantheonStopConfig) -> float:
    """Mirror of ``InitialStop`` — StopTicks below entry (long) / above (short)."""
    return round_to_tick(entry_price - direction * cfg.stop_ticks * cfg.tick_size, cfg.tick_size)


def uses_target(cfg: PantheonStopConfig) -> bool:
    """Only FixedRR carries a target (``UsesTarget`` / ShouldUseTarget)."""
    return cfg.policy == PantheonStopPolicy.FIXED_RR


def initial_target(direction: int, entry_price: float, cfg: PantheonStopConfig) -> float:
    return round_to_tick(entry_price + direction * cfg.target_ticks * cfg.tick_size, cfg.tick_size)


def improves(direction: int, proposed_stop: float, last_stop: float, tick_size: float) -> bool:
    """Ratchet test — long must rise > 0.5 tick, short must fall the same (``Improves``)."""
    if direction == LONG:
        return proposed_stop > last_stop + tick_size * 0.5
    return proposed_stop < last_stop - tick_size * 0.5


def open_profit_currency(direction: int, price: float, entry_price: float,
                         point_value: float, quantity: int) -> float:
    if quantity <= 0:
        return 0.0
    points = price - entry_price if direction == LONG else entry_price - price
    return points * point_value * quantity


def stop_price_from_protected_profit(direction: int, entry_price: float,
                                     protected_profit_currency: float,
                                     point_value: float, quantity: int) -> float:
    """Mirror of ``StopPriceFromProtectedProfit``."""
    denom = point_value * max(1, quantity)
    protected_points = 0.0 if denom == 0.0 else protected_profit_currency / denom
    return entry_price + protected_points if direction == LONG else entry_price - protected_points


def compute_trail_stop(
    direction: int,
    entry_price: float,
    trigger_price: float,
    favorable_extreme: float,
    current_atr: float,
    peak_open_profit_currency: float,
    quantity: int,
    chandelier_ref_price: float,
    breakeven_active: bool,
    cfg: PantheonStopConfig,
) -> TrailResult:
    """1:1 port of ``PantheonStopEngine.ComputeTrailStop`` — the per-policy switch
    from PantheonMaster.ManageLiveDynamicStop. Pure: the caller threads the running
    state and applies the ratchet + side-of-market guard before submitting."""
    tick = cfg.tick_size
    policy = cfg.policy

    if policy == PantheonStopPolicy.BREAK_EVEN_ONLY:
        if breakeven_active:
            return TrailResult()
        trigger_level = entry_price + direction * cfg.breakeven_trigger_ticks * tick
        armed = trigger_price >= trigger_level if direction == LONG else trigger_price <= trigger_level
        if armed:
            return TrailResult(
                has_proposal=True,
                proposed_stop=round_to_tick(entry_price + direction * cfg.breakeven_plus_ticks * tick, tick),
                breakeven_activated=True,
            )
        return TrailResult()

    if policy == PantheonStopPolicy.ATR_TRAIL:
        proposed = favorable_extreme - direction * cfg.atr_trail_multiple * current_atr
        return TrailResult(has_proposal=True, proposed_stop=round_to_tick(proposed, tick))

    if policy == PantheonStopPolicy.CHANDELIER:
        ref_price = (max(chandelier_ref_price, favorable_extreme) if direction == LONG
                     else min(chandelier_ref_price, favorable_extreme))
        proposed = ref_price - direction * cfg.atr_trail_multiple * current_atr
        return TrailResult(has_proposal=True, proposed_stop=round_to_tick(proposed, tick))

    if policy == PantheonStopPolicy.GIVEBACK:
        if peak_open_profit_currency > 0.0:
            protected_profit = peak_open_profit_currency * (1.0 - cfg.giveback_pct)
            proposed = stop_price_from_protected_profit(
                direction, entry_price, protected_profit, cfg.point_value, quantity)
            return TrailResult(has_proposal=True, proposed_stop=round_to_tick(proposed, tick))
        return TrailResult()

    # FixedRR / FixedStop: the initial stop never trails.
    return TrailResult()
